_posts/2025-11-12-adaptive-refresh-for-resilient-segment-replication.md

```diff
@@ -0,0 +1,212 @@
+
+Keen readers would note however, that Lucene’s `refreshIfNeeded` API only refreshed on the latest commit by default. We [contributed changes](https://github.com/apache/lucene/pull/14443) to Lucene, that allow `SearcherManager` (and related classes) to intelligently select the candidate commit for refresh. Consumer implementations can add logic to select the right commit, like calculating the bytes difference between current searcher commit and available commits, and selecting the newest commit that fits in the available searcher memory. Replicas can thus *adaptively refresh* on the bite sized commits. If all commits are small enough, they can directly jump to the latest one. At the other extreme, they can incrementally step through all provided commits. These changes are available as part of Lucene 10.3.
+
+
+### Setting up indexers for bite sized commits
+
+On the indexer side, Lucene's IndexWriter continues to operate as before, creating commits on the indexer machine, though you may want to consider increasing commit frequency to keep each commit as a smaller differential from the previous one. For example, instead of creating a commit every minute, you can consider creating one every 15 seconds, giving you 4 incremental hops per minute window that a replica can chose from. Note that this is one of many possible approaches. Another, better way is to accurately trigger commits whenever segment turnover exceeds a configured threshold size.
+
+`IndexWriters` are configured with an `IndexDeletionPolicy` to clean up older commits that are no longer needed. The default is `KeepOnlyLastCommitDeletionPolicy`, which only retains the most recent commit. Since we want to hold multiple commits in our checkpoint, we must use a different index deletion policy. Lucene offers `KeepLastNCommitsDeletionPolicy` to keep last N commits. You can also write your own deletion policy based on specific business requirements, like keeping all commits within a time window.
+
+Finally, your replication checkpoint should be updated to include all the commits you want bundled together — typically all commits currently available in the index, since your configured deletion policy will already handle cleanup of commits that fall outside the retention window.
+
+With this foundation, we set up our indexers to publish multi-commit checkpoints, enabling replicas to make intelligent, adaptive refresh decisions.
+
+
+### Changes on replicas for adaptive refresh
+
+The replica now receives a set of commits with each checkpoint. Our goal is still to finally refresh on the latest commit. However, instead of directly jumping to the latest commit, replica can now plan its path through a series of commits that are safe to refresh on.
```

Comment:
```suggestion
The replica now receives a set of commits with each checkpoint. Our goal is still to finally refresh on the latest commit. However, instead of directly jumping to the latest commit, the replica can now plan its path through a series of commits that are safe to refresh on.
```